lib/template-loader.php

```diff
@@ -42,34 +42,61 @@ function gutenberg_add_template_loader_filters() {
 	foreach ( $template_types as $template_type ) {
 		add_filter( $template_type . '_template', 'gutenberg_override_query_template', 20, 3 );
 	}
-
-	add_filter( 'template_include', 'gutenberg_template_include_filter', 20 );
 }
 add_action( 'wp_loaded', 'gutenberg_add_template_loader_filters' );
 
 /**
- * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ * Get the template hierarchy for a given template type.
  *
- * The method returns an empty result for every template so that a 'wp_template' post
- * is used instead.
+ * Internally, this filters into the "{$type}_template_hierarchy" hook to record the type-specific template hierarchy.
  *
- * @see gutenberg_template_include_filter
+ * @param string $template_type A template type.
+ * @return string[] A list of template candidates, in descending order of priority.
+ */
+function get_template_hierachy( $template_type ) {
+	$get_template_function     = 'get_' . str_replace( '-', '_', $template_type ) . '_template'; // front-page -> get_front_page_template
+	$template_hierarchy_filter = str_replace( '-', '', $template_type ) . '_template_hierarchy'; // front-page -> frontpage_template_hierarchy
+
+	$result                             = array();
+	$template_hierarchy_filter_function = function( $templates ) use ( &$result ) {
+		$result = $templates;
+		return $templates;
+	};
+
+	add_filter( $template_hierarchy_filter, $template_hierarchy_filter_function, 20, 1 );
+	call_user_func( $get_template_function ); // This invokes template_hierarchy_filter
```

Comment:
Do we need some extra validation here?

Reply:
To make sure that `$get_template_function` really exists? We're building these from a hardcoded list, so I think we should be good? But maybe I'm misunderstanding :sweat_smile:

Reply:
Yes, the static list is not collocated, and people could pass whatever to `get_template_hierachy` in practical use. It'd make sense for me for this function to take its own precautions and not assume the parameter will always come from a static list.

Reply:
:facepalm: Of course. For whatever reason, I was confusing this with the loop over `$template_types` in `lib/edit-site-page.php` (where no additional validation would be required).

Reply:
9a89d4074e